scalameter-core/src/main/scala/org/scalameter/InvocationCountMatcher.scala

```diff
@@ -0,0 +1,123 @@
+   */
+  def matches(methodName: String, methodDescriptor: String): Boolean
+}
+
+/** Matches method with a method name given as a string. */
+case class StringMethodMatcher(method: String) extends MethodMatcher {
+  def matches(methodName: String, methodDescriptor: String): Boolean = methodName == method
+}
+
+/** Matches method with a regex. */
+case class RegexMethodMatcher(regex: Pattern) extends MethodMatcher {
+  def matches(methodName: String, methodDescriptor: String): Boolean = regex.matcher(methodName).matches()
+}
+
+/** Matches method with a [[java.lang.reflect.Method]].
+ *
+ *  That means that method name, method arguments and method return type are matched.
+ */
+case class FullMethodMatcher(method: Method) extends MethodMatcher {
+  def matches(methodName: String, methodDescriptor: String): Boolean =
+    methodName == method.getName && methodDescriptor == Type.getMethodDescriptor(method)
+}
+
+
+/** Pattern used for matching methods whose invocations should be counted.
+ *
+ *  @param classMatcher matches full class name given in internal format ('/' instead of a '.' as a separator)
+ *  @param methodMatcher matches method name
+ */
+case class InvocationCountMatcher(classMatcher: ClassMatcher, methodMatcher: MethodMatcher) {
+  /** Matches allocations of a class. */
+  def this(clazz: Class[_]) = this(StringClassMatcher(clazz), StringMethodMatcher("<init>"))
+
+  /** Matches method name in a class. */
+  def this(clazz: Class[_], methodName: String) = this(StringClassMatcher(clazz), StringMethodMatcher(methodName))
+
+  /** Matches method name with a regex in a class. */
+  def this(clazz: Class[_], methodRegex: Pattern) = this(StringClassMatcher(clazz), RegexMethodMatcher(methodRegex))
+
+  /** Matches method with a [[java.lang.reflect.Method]] in a class. */
+  def this(clazz: Class[_], method: Method) = this(StringClassMatcher(clazz), FullMethodMatcher(method))
+
+  /** Matches allocations of classes matched by a regex. */
+  def this(classRegex: Pattern) = this(RegexClassMatcher(classRegex), StringMethodMatcher("<init>"))
+
+  /** Matches method name in classes matched by a regex. */
+  def this(classRegex: Pattern, methodName: String) = this(RegexClassMatcher(classRegex), StringMethodMatcher(methodName))
+
+  /** Matches method name with a regex in classes matched by a regex. */
+  def this(classRegex: Pattern, methodRegex: Pattern) = this(RegexClassMatcher(classRegex), RegexMethodMatcher(methodRegex))
+
+  /** Matches method with a [[java.lang.reflect.Method]] in classes matched by a regex. */
+  def this(classRegex: Pattern, method: Method) = this(RegexClassMatcher(classRegex), FullMethodMatcher(method))
+  
+  def classMatches(className: String): Boolean = classMatcher.matches(className)
+
+  def methodMatches(methodName: String, methodDescriptor: String): Boolean = methodMatcher.matches(methodName, methodDescriptor)
+}
+
+object InvocationCountMatcher {
```

Comment:
I would leave only `Class`+`Method`, *classname*+*methodName*, *regex*+*regex* and the primary ctor.

Avoid `apply` overloads - keep an `apply` only for the primary ctor.
Name the rest appropriately.

The single arg allocation matcher ctor is surprising - introduce a special factory method `allocations` for that one.